sympy/tensor/array/sparse_ndim_array.py

```diff
@@ -109,8 +109,42 @@ def reshape(self, *newshape):
         if new_total_size != self._loop_size:
             raise ValueError("Invalid reshape parameters " + newshape)
 
-        return type(self)(*(newshape + (self._array,)))
-
+        return type(self)(*(newshape + (self._sparse_array,)))
+
+    def _eval_derivative_array(self, arg):
+        from sympy.core.compatibility import Iterable
+        from sympy.matrices.common import MatrixCommon
+        from sympy import Derivative
+        from sympy import Dict
+        if isinstance(arg, (Iterable, Tuple, MatrixCommon, NDimArray)):
+            return type(self)([self.diff(x) for x in arg])
+        else:
+            s_a = self._sparse_array
+            s_a2 = {}
+            for key in s_a:
+                value = _sympify(s_a[key])
+                deri = Derivative(value, arg, **{'evaluate':True})
+                if deri != 0:
+                    s_a2[key] = deri
+            return type(self)(Dict(s_a2), self._shape)
+
+    def __eq__(self, other):
```

Comment:
Since the evaluation of derivative triggers `__eq__`, this issue is handled together in this PR. I don't know if it is appropriate to do so. Waiting for some advices. :-)

Reply:
What about temporarily ignore equality checks?

Reply:
@Upabjojr Do you mean that we keep equality checks casting sparse array to dense array and handle the case of `__eq__` in another PR?

Reply:
There is already a method handling __eq__ in the superclass. I think the code should remain there.